[WebGPU] Support module constants in vertex and fragment shaders

mwyrzykowski opened the pull request:
#### c4287bb1ee99f7e34944ec7e8a962df6e452bbc9
<pre>
[WebGPU] Support module constants in vertex and fragment shaders
<a href="https://bugs.webkit.org/show_bug.cgi?id=249793">https://bugs.webkit.org/show_bug.cgi?id=249793</a>
&lt;radar://103490403&gt;

Reviewed by Myles C. Maxfield.

There are multiple ways to support module constants. One way
is to use MTLFunctionConstantValues which allows for shader
specialization after the shader is already compiled, however
it does require invoking the compiling again.

This is what this patch implements.

Alternatively we could defer compilation to render pipeline
creation, but that has the drawback of more compilations when
many pipeline objects are created.

* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::validateRenderPipeline):
(WebGPU::buildKeyValueReplacements):
(WebGPU::Device::createRenderPipeline):
Build a hash map containing the names of the constants and their values.

* Source/WebGPU/WebGPU/ShaderModule.h:
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::ShaderModule::getNamedFunction const):
Support named function constants. In the future, we need to get the correct
type information from the compiler instead of assuming the type is float.

* Websites/webkit.org/demos/webgpu/scripts/textured-cube-shader-constants.js: Added.
(async helloCube.frameUpdate):
(async helloCube):
* Websites/webkit.org/demos/webgpu/textured-cube-shader-constants.html: Added.
Add a test which illustrates in the behavior.

Canonical link: <a href="https://commits.webkit.org/258808@main">https://commits.webkit.org/258808@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7960b22a8af8cc25ebbfd6b15e9a3ec9f114767a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102959 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12079 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35981 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112208 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13099 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2983 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95198 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/110169 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108731 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10061 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/93251 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37697 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24792 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79427 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5516 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26201 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5670 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2656 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11682 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45699 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6063 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7420 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/RenderPipeline.mm:
```diff
@@ -333,6 +322,17 @@ static MTLVertexStepFunction stepFunction(WGPUVertexStepMode stepMode)
     return vertexDescriptor;
 }
 
+static auto buildKeyValueReplacements(const auto& stage)
```
I used `auto` here because `descriptor.vertex` and `fragmentDescriptor` are different types

litherum commented on Source/WebGPU/WebGPU/RenderPipeline.mm:
```diff
@@ -333,6 +319,17 @@ static MTLVertexStepFunction stepFunction(WGPUVertexStepMode stepMode)
 
+static auto buildKeyValueReplacements(const auto& stage)
+{
+    HashMap<String, float> keyValueReplacements;
```
WebGPU.h indicates this should be a double, not a float

litherum commented on Source/WebGPU/WebGPU/ShaderModule.mm:
```diff
@@ -239,9 +239,24 @@ auto compilationMessageData(convertMessages(
+    MTLFunctionConstantValues *constantValues = [MTLFunctionConstantValues new];
+    for (auto& kvp : keyValueReplacements) {
+        // FIXME: specify actual types from the compiler information
+        [constantValues setConstantValue:&kvp.value type:MTLDataTypeFloat withName:kvp.key];
```
I think we should fully implement this, or add FIXMEs with bug links. There are 2 pieces missing:

1. As the comment suggests, MTLDataTypeFloat is not the only type. Ideally, this would be done by the WGSL compiler emitting metadata telling you about, for each entry point, which pipeline-overridable constants exist, and what their types are. I think we should make sure that the WGSL compiler's API signatures support returning this data, even if compiler right now can't actually populate that data structure and it has to be shimmed some other way (possibly MTLReflection). We should hook this data structure up here, rather than using a `HashMap<String, float>` that is not expressive enough.
2. Pipeline-overridable constants in WGSL can be identified in 2 ways: either by name or by index. I think we should hook that up while we're here.

mwyrzykowski replied:
I think the second point requires information from the compiler, to tell us the id which corresponds to a given name in the generated metal source as metal has no concept of an id and the ids are not linearly ordered. From the specification, it appears the IDs are basically secondary names:
```
@id(0)      override has_point_light: bool = true;  // Algorithmic control.
@id(1200)   override specular_param: f32 = 2.3;     // Numeric control.
@id(1300)   override gain: f32;                     // Must be overridden.
```

and then `buildKeyValueReplacements(const auto& stage)` could be written like so:

```
    for (size_t i = 0; i < stage.constantCount; ++i) {
         auto& kvp = stage.constants[I];
         auto& remappedKey = kvp.key;
         auto it = idRemappingTable.find(remappedKey);
         if (it != idRemappingTable.end())
              remappedKey = it->value;
         keyValueReplacements.set(String::fromUTF8(remappedKey), kvp.value);
     }
```
or something like that

But I could make an empty `idRemappingTable` so at least the logic is there.

mwyrzykowski replied:
I filed https://bugs.webkit.org/show_bug.cgi?id=250444 for the first issue and will add a FIXME, but we can also do this in the mean time by calling `-[MTLLibrary newFunctionWithName:]` twice and using `-[MTLFunction functionConstantsDictionary]`

litherum commented on Source/WebGPU/WebGPU/ShaderModule.mm:
```diff
@@ -239,9 +239,24 @@ auto compilationMessageData(convertMessages(
+id<MTLFunction> ShaderModule::getNamedFunction(const String& name, const HashMap<String, float>& keyValueReplacements) const
 {
-    return [m_library newFunctionWithName:name];
+    if (!keyValueReplacements.size())
```
What about validation? What happens if there are some pipeline-overridable constants which must be filled in, but they aren't?

mwyrzykowski replied:
per the [Metal specification](https://developer.apple.com/documentation/metal/mtllibrary/1515524-newfunctionwithname) we should return an invalid pipeline if we try to use a MTLFunction to create a render pipeline and don't specify the function constants:

```
    id<MTLRenderPipelineState> renderPipelineState = [m_device newRenderPipelineStateWithDescriptor:mtlRenderPipelineDescriptor options:MTLPipelineOptionArgumentInfo reflection:&reflection error:nil];
    if (!renderPipelineState)
        return RenderPipeline::createInvalid(*this);
```

or if we don't specify some of them, then `-[MTLLibrary newFunctionWithName:constantValues: error:]`  would return nil and we and invalid pipeline when `getNamedFunction` returns.

https://developer.apple.com/documentation/metal/mtllibrary/1640020-newfunctionwithname

We can perform our own validation on top of the validation the Metal API does but it is probably not necessary?

litherum commented on Source/WebGPU/WebGPU/ShaderModule.mm:
```diff
@@ -239,9 +239,24 @@ auto compilationMessageData(convertMessages(
+    id<MTLFunction> result = [m_library newFunctionWithName:name constantValues:constantValues error:&error];
+
+    if (error)
+        WTFLogAlways("MSL compilation error: %@", error);
```
I think we should do this properly (or add a FIXME and a bug link). This is supposed to push an uncategorized error to the current error scope.

mwyrzykowski replied:
I'll add a bug link for proper logging throughout WebGPU.framework

mwyrzykowski committed:
[WebGPU] Support module constants in vertex and fragment shaders
https://bugs.webkit.org/show_bug.cgi?id=249793
<radar://103490403>

Reviewed by Myles C. Maxfield.

There are multiple ways to support module constants. One way
is to use MTLFunctionConstantValues which allows for shader
specialization after the shader is already compiled, however
it does require invoking the compiling again.

This is what this patch implements.

Alternatively we could defer compilation to render pipeline
creation, but that has the drawback of more compilations when
many pipeline objects are created.

* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::validateRenderPipeline):
(WebGPU::buildKeyValueReplacements):
(WebGPU::Device::createRenderPipeline):
Build a hash map containing the names of the constants and their values.

* Source/WebGPU/WebGPU/ShaderModule.h:
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::ShaderModule::getNamedFunction const):
Support named function constants. In the future, we need to get the correct
type information from the compiler instead of assuming the type is float.

* Websites/webkit.org/demos/webgpu/scripts/textured-cube-shader-constants.js: Added.
(async helloCube.frameUpdate):
(async helloCube):
* Websites/webkit.org/demos/webgpu/textured-cube-shader-constants.html: Added.
Add a test which illustrates in the behavior.

Canonical link: https://commits.webkit.org/258808@main
